### ISSUE TITLE
"Return type hint can be used" inspection not working for me

### ISSUE THREAD

**Issue by mfn** [kalessil/phpinspectionsea#574]
Using:
- PhpStorm 2017.3 EAP, Build #PS-173.2941.8, built on October 4, 2017
- Php Inspetions (EA Extended) 2.3.10.6

I've this feature activated:
![image](https://user-images.githubusercontent.com/87493/31330783-1356341e-ace0-11e7-8f3a-3f1fb8a043ba.png)

Language Level is set to 7.1

It definitely used to work but doesn't anymore; In the following screenshot I expect the `testCallbackCalled` to be highlighted with a warning, but it doesn't work:
![image](https://user-images.githubusercontent.com/87493/31330850-59e5d0f6-ace0-11e7-8fc9-9ed056f1d390.png)

(I also tried with the default "Weak warning", doesn't work either).

**Comment by kalessil** [comment]
It probably sees multiple return types or couldn't resolve some of the return types. Difficult to say from this screenshot.

**Comment by mfn** [comment]
Sorry for being so unspecific. What I'm trying to say, it doesn't work at all. This was just a random example I picked up. Same when the method is empty:
![image](https://user-images.githubusercontent.com/87493/31333494-99558bfa-acea-11e7-83ec-660d08a79404.png)

**Comment by kalessil** [comment]
Ok, I see now. Will check if the problem is in EAP.

**Comment by kalessil** [comment]
@mfn : weird stuff, with 2017.3 EAP it seems work
![image](https://user-images.githubusercontent.com/1577185/31499224-7c5b2718-af64-11e7-810b-3fb9174e3e83.png)

I could build a debug version of the plugin which will report types it's resolved (so methods are always reported: regular message or information about resolved types, at least to guess where is an issue)

**Comment by mfn** [comment]
> I could build a debug version 

Before you have to go the extra mile, I will to some testing with a clean repository first and report back. I should find time on the weekend.

**Comment by kalessil** [comment]
@mfn : perhaps it's also due to  #581 (EAP-specific)

**Comment by kalessil** [comment]
I pushed a fresh binary with the mentioned fix

**Comment by mfn** [comment]
In the meantime there was another EAP update, so right now I'm using PS-173.3188.26, built on October 11, 2017 , then I saw these traces for the first time; the look very similar/same to the one from the other report:
```
com.jetbrains.php.lang.psi.elements.Function.getReturnType()Lcom/intellij/psi/PsiElement;
java.lang.NoSuchMethodError: com.jetbrains.php.lang.psi.elements.Function.getReturnType()Lcom/intellij/psi/PsiElement;
	at com.kalessil.phpStorm.phpInspectionsEA.inspectors.codeStyle.UnnecessaryCastingInspector$1.resolveStrictly(UnnecessaryCastingInspector.java:92)
	at com.kalessil.phpStorm.phpInspectionsEA.inspectors.codeStyle.UnnecessaryCastingInspector$1.visitPhpUnaryExpression(UnnecessaryCastingInspector.java:60)
	at com.jetbrains.php.lang.psi.elements.impl.UnaryExpressionImpl.accept(UnaryExpressionImpl.java:27)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:82)
	at com.intellij.codeInspection.InspectionEngine.createVisitorAndAcceptElements(InspectionEngine.java:70)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:285)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:253)
	at com.intellij.concurrency.ApplierCompleter.d(ApplierCompleter.java:133)
	at com.intellij.concurrency.ApplierCompleter.c(ApplierCompleter.java:223)
	at com.intellij.concurrency.ApplierCompleter.d(ApplierCompleter.java:151)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1132)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:105)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:534)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:479)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:94)
	at com.intellij.concurrency.ApplierCompleter.a(ApplierCompleter.java:116)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:96)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:143)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:234)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:96)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

com.jetbrains.php.lang.psi.elements.Function.getReturnType()Lcom/intellij/psi/PsiElement;
java.lang.NoSuchMethodError: com.jetbrains.php.lang.psi.elements.Function.getReturnType()Lcom/intellij/psi/PsiElement;
	at com.kalessil.phpStorm.phpInspectionsEA.inspectors.codeStyle.UnnecessaryCastingInspector$1.resolveStrictly(UnnecessaryCastingInspector.java:92)
	at com.kalessil.phpStorm.phpInspectionsEA.inspectors.codeStyle.UnnecessaryCastingInspector$1.visitPhpUnaryExpression(UnnecessaryCastingInspector.java:60)
	at com.jetbrains.php.lang.psi.elements.impl.UnaryExpressionImpl.accept(UnaryExpressionImpl.java:27)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:82)
	at com.intellij.codeInspection.InspectionEngine.createVisitorAndAcceptElements(InspectionEngine.java:70)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:285)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:253)
	at com.intellij.concurrency.ApplierCompleter.d(ApplierCompleter.java:133)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1127)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:105)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:534)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:479)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:94)
	at com.intellij.concurrency.ApplierCompleter.a(ApplierCompleter.java:116)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:96)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:143)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:234)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:96)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.pollAndExecCC(ForkJoinPool.java:1190)
	at java.util.concurrent.ForkJoinPool.helpComplete(ForkJoinPool.java:1879)
	at java.util.concurrent.ForkJoinPool.awaitJoin(ForkJoinPool.java:2045)
	at java.util.concurrent.ForkJoinTask.doJoin(ForkJoinTask.java:390)
	at java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:719)
	at java.util.concurrent.ForkJoinPool.invoke(ForkJoinPool.java:2616)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgress(JobLauncherImpl.java:65)
	at com.intellij.concurrency.JobLauncher.invokeConcurrentlyUnderProgress(JobLauncher.java:57)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:256)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:226)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.collectInformationWithProgress(LocalInspectionsPass.java:135)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doCollectInformation(ProgressableTextEditorHighlightingPass.java:83)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:70)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.a(PassExecutorService.java:437)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1132)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.d(PassExecutorService.java:430)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:534)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:479)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:94)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.c(PassExecutorService.java:429)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.b(PassExecutorService.java:405)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:143)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:234)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:403)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:170)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```
and
```
java.lang.NoSuchMethodError: com.jetbrains.php.lang.psi.elements.Function.getReturnType()Lcom/intellij/psi/PsiElement;
	at com.kalessil.phpStorm.phpInspectionsEA.inspectors.codeStyle.UnnecessaryCastingInspector$1.resolveStrictly(UnnecessaryCastingInspector.java:92)
	at com.kalessil.phpStorm.phpInspectionsEA.inspectors.codeStyle.UnnecessaryCastingInspector$1.visitPhpUnaryExpression(UnnecessaryCastingInspector.java:60)
	at com.jetbrains.php.lang.psi.elements.impl.UnaryExpressionImpl.accept(UnaryExpressionImpl.java:27)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:82)
	at com.intellij.codeInspection.InspectionEngine.createVisitorAndAcceptElements(InspectionEngine.java:70)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:285)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:253)
	at com.intellij.concurrency.ApplierCompleter.d(ApplierCompleter.java:133)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1127)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:105)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:534)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:479)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:94)
	at com.intellij.concurrency.ApplierCompleter.a(ApplierCompleter.java:116)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:96)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:143)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:234)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:96)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.pollAndExecCC(ForkJoinPool.java:1190)
	at java.util.concurrent.ForkJoinPool.helpComplete(ForkJoinPool.java:1879)
	at java.util.concurrent.ForkJoinPool.awaitJoin(ForkJoinPool.java:2045)
	at java.util.concurrent.ForkJoinTask.doJoin(ForkJoinTask.java:390)
	at java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:719)
	at java.util.concurrent.ForkJoinPool.invoke(ForkJoinPool.java:2616)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgress(JobLauncherImpl.java:65)
	at com.intellij.concurrency.JobLauncher.invokeConcurrentlyUnderProgress(JobLauncher.java:57)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:256)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:226)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.collectInformationWithProgress(LocalInspectionsPass.java:135)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doCollectInformation(ProgressableTextEditorHighlightingPass.java:83)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:70)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.a(PassExecutorService.java:437)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1132)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.d(PassExecutorService.java:430)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:534)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:479)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:94)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.c(PassExecutorService.java:429)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.b(PassExecutorService.java:405)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:143)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:234)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:403)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:170)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

However that was *before* I installed the linked plugin.

I downloaded the plugin from the linked comment and it appears as  2.3.10.6 .

There are no exceptions but it doesn't work either yet, still the same as I initially wrote 🤷‍♀️

**Comment by JProof** [comment]
Perhaps its the same:
after writing the returntype the gray underscore under the methodname will not leave
![0740](https://user-images.githubusercontent.com/12245927/31587822-8ea7cd96-b1e8-11e7-914d-971ee0b32311.png)

**Comment by kalessil** [comment]
@mfn @JProof : exception goes away, but plugin currently doesn;t see return types declaration in EAP - working on it.

**Comment by iltar** [comment]
Can confirm this error in the latest EAP.
```PhpStorm 2017.3 EAP
Build #PS-173.3188.26, built on October 11, 2017
PhpStorm EAP User
Expiration date: November 10, 2017
JRE: 1.8.0_152-release-1024-b5 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.4.0-62-generic
```
```
com.jetbrains.php.lang.psi.elements.Method.getReturnType()Lcom/intellij/psi/PsiElement;
java.lang.NoSuchMethodError: com.jetbrains.php.lang.psi.elements.Method.getReturnType()Lcom/intellij/psi/PsiElement;
	at com.kalessil.phpStorm.phpInspectionsEA.inspectors.languageConstructions.ReturnTypeCanBeDeclaredInspector$1.visitPhpMethod(ReturnTypeCanBeDeclaredInspector.java:72)
	at com.jetbrains.php.lang.psi.elements.impl.MethodImpl.accept(MethodImpl.java:85)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:82)
	at com.intellij.codeInspection.InspectionEngine.createVisitorAndAcceptElements(InspectionEngine.java:70)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:285)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.a(LocalInspectionsPass.java:253)
	at com.intellij.concurrency.ApplierCompleter.d(ApplierCompleter.java:133)
	at com.intellij.concurrency.ApplierCompleter.c(ApplierCompleter.java:223)
	at com.intellij.concurrency.ApplierCompleter.d(ApplierCompleter.java:151)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1132)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:105)
	at com.intellij.openapi.progress.impl.CoreProgressManager.a(CoreProgressManager.java:534)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:479)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:94)
	at com.intellij.concurrency.ApplierCompleter.a(ApplierCompleter.java:116)
	at com.intellij.concurrency.ApplierCompleter.b(ApplierCompleter.java:96)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:143)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:234)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:96)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by kalessil** [comment]
Issues should be resolved with 2.3.11, which happening tomorrow. I'll close the ticket if the issue remains after the update, we'll re-open it.
